src/visitable.cpp

```diff
@@ -72,14 +72,30 @@ bool visitable<T>::has_item_with( const std::function<bool( const item & )> &fil
     } ) == VisitResponse::ABORT;
 }
 
+/** Sums the two terms, being careful to not trigger overflow.
+ * Doesn't handle underflow.
+ *
+ * @param a The first addend.
+ * @param b The second addend.
+ * @return the sum of the addends, but truncated to std::numeric_limits<int>::max().
+ */
+template <typename T>
+static T sum_no_wrap( T a, T b ) {
+    if( a > std::numeric_limits<T>::max() - b ||
+        b > std::numeric_limits<T>::max() - a ) {
+        return std::numeric_limits<T>::max();
+    }
+    return a + b;
+}
+
 template <typename T>
 static int has_quality_internal( const T& self, const quality_id &qual, int level, int limit )
 {
     int qty = 0;
 
     self.visit_items( [&qual, level, &limit, &qty]( const item *e ) {
         if( e->get_quality( qual ) >= level ) {
-            qty += e->count_by_charges() ? e->charges : 1;
+            qty = sum_no_wrap( qty, e->count_by_charges() ? int( e->charges ) : 1 );
```

Comment:
If `e->charges` (which is a `long`) has a value that's not representable in an `int`, then you get implementation-defined behaviour if you narrow it at this point. This happens before `sum_no_wrap` is called, so the anti-wrap code in that function does not help - you've already overflowed.

If you sum using `qty` as a `long`, then `sum_no_wrap` avoids any overflow correctly, you can still safely compare to `limit` (it will be promoted to `long` for the comparison) and you can safely use `std::min` in the return statement for the same reason (it will be `std::min<long>`). The returned value of `std::min` is guaranteed to be representable as an `int` since `limit` is an int, so narrowing as the final step is safe.

For a concrete example, assume `INT_MAX` is `2**31`, `limit` is 1000, and `e->charges` is `2**40`. You should return 1000 in this case.